core/src/software/aws/toolkits/core/credentials/sso/SsoAccessTokenProvider.kt

```
@@ -0,0 +1,135 @@
+        }
+
+        // Based on botocore: https://github.com/boto/botocore/blob/5dc8ee27415dc97cfff75b5bcfa66d410424e665/botocore/utils.py#L1769
+        val registerResponse = client.registerClient {
+            it.clientType(CLIENT_REGISTRATION_TYPE)
+            it.clientName("aws-toolkit-jetbrains-${Instant.now(clock)}")
+        }
+
+        val registeredClient = ClientRegistration(
+            registerResponse.clientId(),
+            registerResponse.clientSecret(),
+            Instant.ofEpochSecond(registerResponse.clientSecretExpiresAt())
+        )
+
+        cache.saveClientRegistration(ssoRegion, registeredClient)
+
+        return registeredClient
+    }
+
+    private fun authorizeClient(clientId: ClientRegistration): Authorization {
+        // Should not be cached, only good for 1 token and short lived
+        val authorizationResponse = try {
+            client.startDeviceAuthorization {
+                it.startUrl(ssoUrl)
+                it.clientId(clientId.clientId)
+                it.clientSecret(clientId.clientSecret)
+            }
+        } catch (e: InvalidClientException) {
+            cache.invalidateClientRegistration(ssoRegion)
+            throw e
+        }
+
+        return Authorization(
+            authorizationResponse.deviceCode(),
+            authorizationResponse.userCode(),
+            authorizationResponse.verificationUri(),
+            authorizationResponse.verificationUriComplete(),
+            Instant.now(clock).plusSeconds(authorizationResponse.expiresIn().toLong()),
+            authorizationResponse.interval()?.toLong()
+                ?: DEFAULT_INTERVAL
+        )
+    }
+
+    private suspend fun pollForToken(): AccessToken {
+        val registration = registerClient()
+        val authorization = authorizeClient(registration)
+
+        onPendingToken.tokenPending(authorization)
+
+        var backOffTime = Duration.ofSeconds(authorization.pollInterval)
+
+        while (true) {
```

do we need a max retry count?

---

Nope, we aren't supposed to have one. We are to respect the expiresIn / go until the service says do not keep trying. They throw `ExpiredTokenException` and we land in the catch all block